parser-typechecker/src/Unison/Codebase/Branch.hs

```diff
@@ -500,9 +508,15 @@ empty0 :: Branch0 m
 empty0 =
   Branch0 mempty mempty mempty mempty mempty mempty mempty mempty mempty mempty
 
--- | Checks whether a Branch0 is empty.
+-- | Checks whether a Branch0 is empty, which means that the branch contains no terms or
+-- types, and that the heads of all children are empty by the same definition.
+-- This is not as easy as checking whether the branch is equal to the `empty0` branch
+-- because child branches may be empty, but still have history.
 isEmpty0 :: Branch0 m -> Bool
-isEmpty0 = (== empty0)
+isEmpty0 (Branch0 _terms _types _children _edits deepTerms deepTypes _deepTermMetadata _deepTypeMetadata _deepPaths deepEdits) =
+  Relation.null deepTerms
+    && Relation.null deepTypes
+    && Map.null deepEdits
```

Comment:
I'd suggest testing `_terms`, `_types` `_children` and `_edits` instead of the `deep*` variants, just because my gut says the latter might go away when we try to simplify `Branch0`.